### PR TITLE
Print webpack console to browser console

### PR DESCRIPTION
**Pull request by Wernerson**
**What kind of change does this PR introduce?**
feature

**Did you add or update the `examples/`?**
yes

**Summary**
As discussed in issue #1032, the webpack console now gets displayed in the browser console if the progress option is set to true.

**Does this PR introduce a breaking change?**
It shouldn't ;)

**Other information**
None

### CODE REVIEW

**Comment by Wernerson** [comment]
There were some merge issues. Will check them this weekend and make the branch ready to merge 👍

**Comment by shellscape** [comment]
I'd also remove the clearing of the console. Best to leave that up to the users to do manually.

**Comment by Wernerson** [comment]
Done and ready to merge 👍

**Comment by shellscape** [comment]
@Wernerson please update the PR description to remove the bit about clearing the console and we'll get this merged.

File: lib/Server.js [hunk 1]
```diff
@@ -52,8 +53,15 @@ function Server(compiler, options) {
   const invalidPlugin = () => {
     this.sockWrite(this.sockets, 'invalid');
   };
+  const progressPlugin = new webpack.ProgressPlugin((percent, msg, addInfo) => {
```
Comment (shellscape) [inline lib/Server.js]: should probably only load this plugin if `options.progress` is true as well

**Comment by Wernerson** [comment]
@shellscape Condition added and description edited. Codecov fails though...

**Comment by shellscape** [comment]
I'm going to bypass the code coverage this time. The refactor for v3 should help boost those numbers, and the config for that needs to ignore client files until we figure out a good way to test the client. Thanks for getting this in.

File: lib/Server.js [hunk 2]
```diff
@@ -52,6 +53,15 @@ function Server(compiler, options) {
   const invalidPlugin = () => {
     this.sockWrite(this.sockets, 'invalid');
   };
+  if (this.progress) {
+    const progressPlugin = new webpack.ProgressPlugin((percent, msg, addInfo) => {
+      percent = Math.floor(percent * 100);
+      if (percent === 100) msg = 'Compilation competed';
```
Comment (sokra) [inline lib/Server.js]: `Compilation competed` -> `Compilation completed`

File: lib/optionsSchema.json [hunk 3]
```diff
@@ -97,6 +97,14 @@
...
       ]
     },
+    "progress": {
+      "description": "Shows compilation progress in browser console.",
+      "anyOf": [
+        {
+          "type": "boolean"
+        }
+      ]
+    },
```
Comment (sokra) [inline lib/optionsSchema.json]:
``` js
"progress": {
  "description": "Shows compilation progress in browser console.",
  "type": "boolean"
 },
```

File: client/index.js [hunk 4]
```diff
@@ -122,6 +123,14 @@ const onSocketMsg = {
       }
     }
   },
+  progress(progress) {
+    if (typeof document !== 'undefined') {
+      useProgress = progress;
+    }
+  },
+  'progress-update': function progressUpdate(data) {
+    if (useProgress) log.info(`[WDS] ${data.percent}% - ${data.msg}.`);
```
Comment (sokra) [inline client/index.js]: Isn't this super spammy?

Comment (shellscape) [reply client/index.js]: given that it's an opt-in and the browser console doesn't have control characters, I don't believe it is.

File: lib/Server.js (hunk 2 above)
Comment (shellscape) [reply lib/Server.js]: I'll clean up that typo in a bit

**Comment by pelotom** [comment]
I was looking forward to this feature but it doesn't seem to be working, what all is needed to enable it? I set `progress: true` in both the dev server config and in the webpack config, but no dice. FWIW I'm using an ejected `create-react-app`.

**Comment by shellscape** [comment]
@pelotom I just checked the example app for progress and it appears to be working for me.

**Comment by Wernerson** [comment]
@pelotom Whats your loglevel set to? Has to be on info at least.

**Comment by g0t4** [comment]
I'd like to use this feature but on fast builds it is super chatty, and so I disabled it and will probably forget about it when the build starts slowing down :)

How about throttle the console messages if they come in rapidly? 

- That way if the build is slow you get a good indication of progress. 
- And if the build is somewhat slow you get fewer messages. 
- And if the build is super fast you maybe see one or two of these or maybe none. 

Or maybe a simple way to hook into this as a user and apply my own filtering?

**Comment by Wernerson** [comment]
@g0t4 Good idea, I'll take a look into it.

**Pull request merged**